Mono.Debugger.Soft/Mono.Debugger.Soft/DelayedLambdaType.cs

```diff
@@ -0,0 +1,147 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text.RegularExpressions;
+
+namespace Mono.Debugger.Soft
+{
+	// Represents a temporary type of a lambda expression.
+	// In order to determine the type of a lambda based on
+	// its context, type evaluation for it must be delayed.
+	public class DelayedLambdaType : Mirror
+	{
+		class ParsedType
+		{
+			public string nameSpace;
+			public string typeName;
+			public string[] argTypeNames;
+		}
+
+		string expression;
+
+		public DelayedLambdaType (VirtualMachine vm, string expression) : base (vm, 0)
+		{
+			this.expression = expression;
+		}
+
+		public string Expression {
+			get { return expression; }
+		}
+
+		public bool IsAcceptableType (TypeMirror t)
+		{
+			return IsAcceptable (t.FullName);
+		}
+
+		private bool IsAcceptable (string fullName)
+		{
+			return IsAcceptable (ParseFullName (fullName));
+		}
+
+		private bool IsAcceptable (ParsedType t)
+		{
+			return t.nameSpace == "System" && t.typeName == "Func";
+		}
+
+		public string GetLiteralType (TypeMirror typ)
+		{
+			return GetLiteralType (typ.FullName);
+		}
+
+		public string GetLiteralType (string fullName)
+		{
+			ParsedType t = ParseFullName (fullName);
+			if (!IsAcceptable (t))
+				return null;
+			string types = String.Join (",", t.argTypeNames);
+			return t.nameSpace + "." + t.typeName + "<" + types + ">";
+		}
+
+		private static ParsedType ParseFullName (TypeMirror t)
+		{
+			return ParseFullName (t.FullName);
+		}
+
+		private static ParsedType ParseFullName (string fullName)
+		{
+			try {
+				return OnParseFullName (fullName);
+			} catch (Exception) {
+				return null;
+			}
+		}
+
+		private static ParsedType OnParseFullName (string fullName)
+		{
+			// Parse fullName of type to ParsedType that is defined above. Examples for the parsing
+			// are following.
+			// 1) System.Action`1[[System.Int32, mscorlib, Version=xxx, Culture=xxx,
+			//    PublicKeyToken=xxx]]
+			//  => { nameSpace: "System", typeName: "Action", argTypeNames: ["System.Int32"] }
+			// 2) System.Func`2[[System.Int32, mscorlib, Version=xxx, Culture=xxx,
+			//    PublicKeyToken=xxx],[System.Single, mscorlib, Version=xxx, Culture=xxx,
+			//    PublicKeyToken=xxx]]
+			//  => { nameSpace: "System", typeName: "Func",
+			//       argTypeNames: ["System.Int32", "System.Single"] }
+
+			string nameSpace = "";
+			string typeName = "";
+			string rest = fullName;
+			if (rest.StartsWith ("System.", StringComparison.Ordinal)) {
+				nameSpace = "System";
+				rest = rest.Substring (7);
+			} else {
+				throw new ArgumentException ("Currently, we support only lambda whose namespace is `System`");
```

Comment:
I set the limitation because this PR only supports System.Func typed lambda. My local version has already supported other lambdas, so this check was removed.